odbc/src/api.rs

```diff
@@ -695,16 +861,108 @@ pub extern "C" fn SQLGetDiagFieldW(
+                        buffer_length as usize,
+                        text_length_ptr,
+                    ),
+                    None => return SqlReturn::NO_DATA,
+                }
+            }
+            None => return SqlReturn::INVALID_HANDLE,
+        },
+        HandleType::Dbc => match unsafe { (*mongo_handle).as_connection() } {
+            Some(dbc) => {
+                let dbc_contents = (*dbc).read().unwrap();
+                match dbc_contents.sql_states.get(rec_number) {
+                    Some(sql_state) => set_sql_state(sql_state.clone(), state),
+                    None => return SqlReturn::NO_DATA,
+                }
+
+                match dbc_contents.error_messages.get(rec_number) {
+                    Some(error_message) => set_error_message(
+                        error_message.clone(),
+                        message_text,
+                        buffer_length as usize,
+                        text_length_ptr,
+                    ),
+                    None => return SqlReturn::NO_DATA,
+                }
+            }
+            None => return SqlReturn::INVALID_HANDLE,
+        },
+        HandleType::Stmt => match unsafe { (*mongo_handle).as_statement() } {
+            Some(stmt) => {
+                let stmt_contents = (*stmt).read().unwrap();
+
+                match stmt_contents.sql_states.get(rec_number) {
+                    Some(sql_state) => set_sql_state(sql_state.clone(), state),
+                    None => return SqlReturn::NO_DATA,
+                }
+
+                match stmt_contents.error_messages.get(rec_number) {
+                    Some(error_message) => set_error_message(
+                        error_message.clone(),
+                        message_text,
+                        buffer_length as usize,
+                        text_length_ptr,
+                    ),
+                    None => return SqlReturn::NO_DATA,
+                }
+            }
+            None => return SqlReturn::INVALID_HANDLE,
+        },
+        HandleType::Desc => match unsafe { (*mongo_handle).as_descriptor() } {
+            Some(desc) => {
+                let desc_contents = (*desc).read().unwrap();
+                match desc_contents.sql_states.get(rec_number) {
+                    Some(sql_state) => set_sql_state(sql_state.clone(), state),
+                    None => return SqlReturn::NO_DATA,
+                }
+
+                match desc_contents.error_messages.get(rec_number) {
+                    Some(error_message) => set_error_message(
+                        error_message.clone(),
+                        message_text,
+                        buffer_length as usize,
+                        text_length_ptr,
+                    ),
+                    None => return SqlReturn::NO_DATA,
+                }
+            }
+            None => return SqlReturn::INVALID_HANDLE,
+        },
+    }
+    SqlReturn::SUCCESS
 }
 
 #[no_mangle]
```

Comment:
Do we also want to implement the corresponding unicode function `SQLGetDiagRecW()` (next line)?

Reply:
I think we should focus on the W methods (unicode) first because this are the methods PowerBI uses.
As a general rule, the W and not W methods should be exactly the same (extract the code in another function?) except for the C to Rust (input and output buffers) conversions.

Reply:
I see -- I'll move this code into `SQLGetDiagRecW` then!

Reply:
@nbagnard yes, in fact, I would say the extracted code could go in the core API while the unsafe ODBC part will convert the char/wchar to rust string and vice versa where necessary.